### PR TITLE
[AIRFLOW-4766] Add autoscaling option to DataprocClusterCreateOperator

### PR DESCRIPTION
**Pull request by nuclearpinguin**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

File: airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+"""
+Example Airflow DAG that creates DataProc cluster.
+"""
+
+import airflow
+from airflow import models
+from airflow.contrib.operators.dataproc_operator import DataprocClusterCreateOperator
+
+default_args = {"start_date": airflow.utils.dates.days_ago(1)}
```
Comment (mik-laj) [inline airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]: I prefer to keep this in one line with constructor.

File: airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+cluster_name = "testcluster-800"
```
Comment (mik-laj) [inline airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]:
```suggestion
CLUSTER_NAME = "testcluster-800"
```
Constant should use upper case letter.

File: airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+zone = ""
+
+# When using autoscaling the number of primary workers
+# must be within autoscaler min/max range
+num_workers = 2
+
+# Autoscaling policy
+# The policy must be in the same project and Cloud Dataproc region
+# but the zone doesn't have to be the same
+scaling_policy = 'test-policy'
```
Comment (mik-laj) [inline airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]:
```suggestion
SCALING_POLICY_ID = 'test-policy'
```
It is very important here to mark that this is not a concrete policy, but an ID only.

File: airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+policy_uri = 'projects/{p}/locations/{r}/autoscalingPolicies/{id}'.format(
```
Comment (mik-laj) [inline airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]:
```suggestion
POLICY_NAME = 'projects/{project_id}/locations/{region}/autoscalingPolicies/{autoscaling_policy_id}'.format(
```
I think it's worth creating a static method in the hook from this code snippet.

File: airflow/contrib/operators/dataproc_operator.py [hunk 5]
```diff
@@ -99,6 +99,10 @@ class DataprocClusterCreateOperator(DataprocOperationBaseOperator):
     :param custom_image: custom Dataproc image for more info see
         https://cloud.google.com/dataproc/docs/guides/dataproc-images
     :type custom_image: str
+    :param autoscaling_policy: The autoscaling policy used by the cluster. Only resource names
+        including projectid and location (region) are valid. Example:
+        projects/[projectId]/locations/[dataproc_region]/autoscalingPolicies/[policy_id]
```
Comment (mik-laj) [inline airflow/contrib/operators/dataproc_operator.py]:
```suggestion
        ``projects/[projectId]/locations/[dataproc_region]/autoscalingPolicies/[policy_id]``
```

File: tests/contrib/operators/test_dataproc_operator.py [hunk 6]
```diff
@@ -119,6 +120,7 @@ def setUp(self):
                     project_id=GCP_PROJECT_ID,
                     num_workers=NUM_WORKERS,
                     zone=GCE_ZONE,
+                    autoscaling_policy=SCALING_POLICY,
```
Comment (mik-laj) [inline tests/contrib/operators/test_dataproc_operator.py]: What do you think about system tests? Do you have time to pay the technological debt?

Comment (potiuk) [reply tests/contrib/operators/test_dataproc_operator.py]: I think this is not the righ time to add system tests. I would add them later if we have time.

File: airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py (hunk 2 above)
Comment (potiuk) [reply airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]: Yep. and below.

File: airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py (hunk 3 above)
Comment (potiuk) [reply airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]: Agree.

File: airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py (hunk 4 above)
Comment (potiuk) [reply airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]: For sure the short variable names are too short :). I prefer readable names (and this is what pylint expects)

Comment (nuclearpinguin) [reply airflow/contrib/example_dags/example_gcp_dataproc_create_cluster.py]: I think that I will remove the autoscaling policy from the example for now. It should be added with proper operator (`DataprocCreatePolicy` or something) making the example fully functional:
`create_policy >> create_cluster >> delete_cluster`

**Comment by nuclearpinguin** [comment]
@mik-laj, @potiuk should we merge those changes?

**Comment by mik-laj** [comment]
You should create a PR in Apache repository. :-D

**Pull request closed without merging**